File: Terrazas-Chavez-Jose-Alexis/pipes/pipes_2.py
'''
Porgrama que implementa comunicación inter-procesos usando un pipe en Python,
con el propósito de demostrar cómo los procesos pueden intercambiar información 
mediante un canal de comunicación unidireccional.
'''
import os
import time

def writer(pipe_write):
    """Proceso escritor: escribe datos periódicamente en el pipe."""
    for i in range(1, 6):  # Escribe 5 veces
        message = f"Mensaje {i}\n".encode()
        os.write(pipe_write, message)
        print(f"Escritor: escribió '{message.decode().strip()}'")
        time.sleep(2)  # Espera 2 segundos entre escrituras
    os.close(pipe_write)  # Cierra el extremo de escritura del pipe

def reader(pipe_read):
    """Proceso lector: lee datos del pipe. Se bloquea si el pipe está vacío."""
    while True:
        try:
            # Lee del pipe
            data = os.read(pipe_read, 1024)  # Intenta leer hasta 1024 bytes
            if not data:  # Si no hay más datos, rompe el bucle
                break
            print(f"Lector: leyó '{data.decode().strip()}'")
        except OSError:
            print("Error al leer del pipe.")
            break
    os.close(pipe_read)  # Cierra el extremo de lectura del pipe


def main():
    # Crear el pipe
    r_fd, w_fd = os.pipe()

    # Crear procesos para lectura y escritura
    pid = os.fork()  # Dividir en dos procesos

    if pid == 0:  # Proceso hijo: lector
        os.close(w_fd)  # Cerrar extremo de escritura en el lector
        reader(r_fd)
    else:  # Proceso padre: escritor
        os.close(r_fd)  # Cerrar extremo de lectura en el escritor
        writer(w_fd)
        os.wait()  # Esperar a que el hijo termine

if __name__ == "__main__":
    main()

File: Perez-Meneses-Juan-Carlos/Pipes/pipes2.java
import java.io.*;

public class PipesExample {
    public static void main(String[] args) {
        try {
            // Crear los pipes
            PipedInputStream inputStream = new PipedInputStream();
            PipedOutputStream outputStream = new PipedOutputStream(inputStream);

            String buf = "abcde";
            int tam = buf.length();

            // Crear un proceso hijo usando Threads (simula fork)
            Thread childProcess = new Thread(() -> {
                try {
                    byte[] buffer = new byte[tam];
                    inputStream.read(buffer, 0, tam);
                    System.out.println("H= " + new String(buffer));
                } catch (IOException e) {
                    System.out.println("Error en el proceso hijo: " + e.getMessage());
                }
            });

            childProcess.start(); // Iniciar el proceso hijo

            // Proceso padre
            Thread.sleep(15000); // Simula sleep(15)
            outputStream.write(buf.getBytes(), 0, tam);
            System.out.println("P= " + buf);

            // Cerrar recursos
            outputStream.close();
            inputStream.close();
            childProcess.join(); // Esperar a que el hijo termine

        } catch (IOException | InterruptedException e) {
            System.out.println("Error general: " + e.getMessage());
        }
    }
}
